File: python/flask/sample/views/open_cades_express.py
import os
import uuid

from flask import Blueprint, current_app, render_template
from pkiexpress import CadesSignatureExplorer

from sample.utils import set_pki_defaults

# 26-08-2022
# By further inspecting in the latest Blueprint documentation (https://flask.palletsprojects.com/en/2.2.x/api/#blueprint-objects), 
# when creating a Blueprint object, the first parameter (name) is prepend to the URL endpoint. Therefore, Blueprint no longer 
# allows dots in the name since it would break the URL entirely.
__name__ = __name__.replace(".","/")
blueprint = Blueprint(os.path.basename(__name__), __name__,
                      url_prefix='/open-cades-express')


@blueprint.route('/<file_id>')
def index(file_id):

    # Verify if the provided userfile exists.
    file_path = os.path.join(current_app.config['APPDATA_FOLDER'], file_id)
    if not os.path.exists(file_path):
        return render_template('error.html', msg='File not found')

    # Get an instance of the CadesSignatureExplorer class, used to open/validate
    # PKCS #12 signatures.
    sig_explorer = CadesSignatureExplorer()

    # Set PKI default options (see utils.py)
    set_pki_defaults(sig_explorer)

    # Specify that we want to validate the signatures in the file, not only
    # inspect them.
    sig_explorer.validate = True

    # Set the file to be inspected.
    sig_explorer.set_signature_file_from_path(file_path)

    output_file = '%s.pdf' % (str(uuid.uuid4()))
    sig_explorer.extract_content_path = os.path.join(current_app.config['APPDATA_FOLDER'], output_file)
    # Call the open() method, which returns the signature file's information.
    signature = sig_explorer.open()

    # Render the signature opening page.
    return render_template('/open_cades_express/index.html',
                           signature=signature, output_file=output_file)

File: nodejs/expressmvc/routes/open-cades-express.js
// - TODO: Uncomment when PKI Express supports cades validation
// const express = require('express');
// const path = require('path');
// const uuidv4 = require('uuid/v4');
// const { CadesSignatureExplorer } = require('pki-express');

// const { Util } = require('../util');
// const { StorageMock } = require('../storage-mock');

// const router = express.Router();
// const APP_ROOT = process.cwd();

// router.get('/', (req, res, next) => {
// 	const { fileId } = req.query;

// 	// Verify if the provided userfile exists
// 	if (!StorageMock.existsSync({ fileId })) {
// 		const notFound = new Error('The fileId was not found');
// 		notFound.status = 404;
// 		next(notFound);
// 		return;
// 	}
// 	const filePath = StorageMock.getDataPath(fileId);

// 	// Get an instance of the CadesSignatureExplorer class, used tod open/validate
// 	// PKCS #12 signatures.
// 	const sigExplorer = new CadesSignatureExplorer();

// 	// Set PKI default options (see utils.js)
// 	Util.setPkiDefaults(sigExplorer);

// 	// Specify that we want to validate the signatures in the file, not only
// 	// inspect them.
// 	sigExplorer.validate = true;

// 	// Specify that we want to validate the signatures in the file, not only
// 	// inspect them.
// 	sigExplorer.setSignatureFileFromPathSync(filePath);

// 	// Generate path for output file and add the signature finisher.
// 	StorageMock.createAppData(); // Make sure the "app-data" folder exists (util.js).
// 	const outputFile = `${uuidv4()}.pdf`;
// 	sigExplorer.setExtractContentPath(path.join(APP_ROOT, 'app-data', outputFile));

// 	// Call the open() method, which returns the signature file's information.
// 	sigExplorer.open()
// 		.then((signature) => {
// 			// Render the signature opening page.
// 			res.render('open-cades-express', { signature, outputFile });
// 		})
// 		.catch((err) => next(err));
// });

// module.exports = router;
